pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -1943,5 +2039,44 @@ func needsSCSIControler(vmi *v1.VirtualMachineInstance) bool {
 			return true
 		}
 	}
-	return false
+	return !vmi.Spec.Domain.Devices.DisableHotplug
+}
+
+func getPrefixFromBus(bus string) string {
+	switch bus {
+	case "virtio":
+		return "vd"
+	case "sata", "scsi":
+		return "sd"
+	case "fdc":
+		return "fd"
+	default:
+		log.Log.Errorf("Unrecognized bus '%s'", bus)
+		return ""
+	}
+}
+
+func newDeviceNamer(volumeStatuses []v1.VolumeStatus, disks []v1.Disk) map[string]deviceNamer {
+	prefixMap := make(map[string]deviceNamer)
+	volumeTargetMap := make(map[string]string)
+	for _, volumeStatus := range volumeStatuses {
+		volumeTargetMap[volumeStatus.Target] = volumeStatus.Name
+	}
```

Comment:
I've been looking at how the target device names are picked/assigned and i think there might be a race condition here.

Here's the issue the best I can describe it.

When we pick the device name, we're setting it at the point that we Attach the disk using libvirt, then we're discovering the name we picked in the virt-handler/vm.go reconcile loop after the fact. 

Now, lets say we add/remove/add/remove lots of disks and volumes right after one another. It seems technically possible if all the timing lines up perfectly that we can get into a situation where we attempt to assign a target device that has technically already been assigned simply because we haven't detected that the device name has been assigned quite yet by the libvirt domain event.

Is what I'm describing possible? If so, should we consider moving the device name assignment outside of the converter code and make sure the dev is stored in each volumeStatus entry before those volumes are actually attached to the libvirt domain?

Reply:
We talked this out, and for others I will summarize the conclusion. Basically this won't be a problem because lets say we do a get a collision, and we attempt to hot plug with a target that is already in use. The attach will fail, causing a sync error on the domain. This will re-queue the sync, this gives us time to update the volume statuses, and eventually it will get updated. Each reconcile loop we build a new map to determine the devices to allocate, based on the values in the volume statuses. Which will now contain the collision, and it will generate a different name and should succeed.